lexicon/providers/safedns.py

```diff
@@ -0,0 +1,200 @@
+        url = '/zones/{0}/records'.format(self.domain_id)
+        records = []
+        payload = {}
+
+        # Handle pagination and iteratively collect all records
+        next_url = url
+        while next_url is not None:
+            payload = self._get(next_url)
+            if 'meta' in payload \
+                    and 'pagination' in payload['meta'] \
+                    and 'links' in payload['meta']['pagination'] \
+                    and 'next' in payload['meta']['pagination']['links']:
+                next_url = payload['meta']['pagination']['links']['next']
+            else:
+                next_url = None
+
+            # Assign the returned attributes to the keys that lexicon expects
+            for record in payload['data']:
+                processed_record = {
+                    'id': record['id'],
+                    'name': record['name'],
+                    'type': record['type'],
+                    # Make sure TXT records are wrapped before sending
+                    'content': self._remove_quotes(record['type'], record['content']),
+                    'updated_at': record['updated_at'],
+                    'ttl': record['ttl'],
+                    'priority': record['priority']
+                }
+                records.append(processed_record)
+
+        # This is filtering logic to return only the record which matches what has
+        # been passed in to the method
+        if rtype:
+            records = [
+                record for record in records
+                if record['type'] == rtype]
+        if name:
+            records = [
+                record for record in records
+                # DNS should not be case-sensitive, so perform lower-case comparison
+                if record['name'].lower() == self._full_name(name).lower()]
+        if content:
+            records = [
+                record for record in records
+                if record['content'].lower() == content.lower()]
+
+        LOGGER.debug('list_records: %s', records)
+        return records
+
+    def _create_record(self, rtype, name, content):
+
+        # Check whether the record already exists with the same  rtype, name & content.
+        # If so, claim to have added the record, but dont't do anything.
+        records = self._list_records(rtype, name, content)
+        if records:
+            LOGGER.debug('create_record: (ignored, duplicate record): %s', records[0]['id'])
+            return True
+
+        # Make sure TXT records are wrapped in quotes
+        if content:
+            content = self._add_quotes(rtype, content)
+
+        data = {
+            'name': self._full_name(name),
+            'type': rtype,
+            'content': content
+        }
+
+        self._post('/zones/{0}/records'.format(self.domain), data)
+        LOGGER.debug('create_record: %s', True)
+        return True
+
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
```

Comment:
There is one case you did not implement here, which is related to the integration test `test_provider_when_calling_update_record_should_modify_record_name_specified`. It is when Lexicon is invoked to update a record, but no identifier is passed. In this case, the `--name` flag must be passed, and `name` and `rtype` are used together to find the record.

I usually reuse ̀self.list_records` for that. If there is no record found, then there is no update. If at least one record is found, the first one is updated (with a warning if there is more than one record).

Once you have implemented that, don't forget to remove `test_provider_when_calling_update_record_should_modify_record_name_specified` from the skipped tests, and fill its associated recording.

Reply:
I saw your logic in the ovh provider and I've mirrored similar in mine. Hopefully it satisfies.